File: frontend/src/services/forYou.ts
// forYou.ts — the three calls the "For you" page makes (CC_TASK_FOR_YOU_v1 L1).
//
// One person's list across every deck in the case, the badge's number, and the
// write that clears a question when she opens it.
//
// ## Every string on the page is the SERVER's
//
// A row arrives with its deck line, its body and its byline already composed,
// and the tab labels arrive with their counts in them. Nothing here fills a
// template, and nothing here decides what a row says — see
// `backend/src/services/for_you_rows.rs` for why.
//
// Every call goes through `authFetch` (its own AbortController timeout) and
// every failure throws a sentence naming what did not happen. Nothing is
// swallowed: the page shows the sentence.

import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";
import { readErrorMessage } from "./fetchUtils";
import { PRACTICE_TIMEOUT_MS } from "./practice";

/** Whose list this viewer is served. Mirrors `ForYouSide`. */
export type ForYouSide = "reviewers" | "witness" | "none";

/** Which heading a row sits under. Decided on the server, in the case's timezone. */
export type ForYouDay = "today" | "yesterday" | "earlier";

/** One row. Checked BY EYE against `ForYouRowDto`. */
export type ForYouRow = {
  kind: string;
  item_id: string;
  scenario_id: string;
  /** Absent for a note about a whole scenario, which opens the deck instead. */
  question_id?: string;
  deck_line: string;
  body: string;
  byline: string;
  when: string;
  day: ForYouDay;
  read: boolean;
};

/**
 * The page's words, as the browser receives them.
 *
 * A map rather than a named type for the reason `PracticeWording` is one: the
 * keys are settings rows and the set moves with the store. `wordingOf` below is
 * the single reader, and it throws BY NAME on a key the payload does not carry
 * — never a silent `undefined` rendered as an empty label.
 */
export type ForYouWording = Record<string, string>;

/** The whole page. Checked by eye against `ForYouPayload`. */
export type ForYouPage = {
  side: ForYouSide;
  subtitle: string;
  tab_unread_label: string;
  tab_everything_label: string;
  unread_count: number;
  unread: ForYouRow[];
  everything: ForYouRow[];
  empty_hint: string;
  /** Absent when this side has never had an item — the line is then not shown. */
  empty_last?: string;
  wording: ForYouWording;
};

/** The badge's number. Checked by eye against `ForYouSummaryDto`. */
export type ForYouSummary = {
  side: ForYouSide;
  unread_count: number;
};

/** One stored string, or a sentence naming the key that is missing. */
export function wordingOf(wording: ForYouWording, key: string): string {
  const value = wording[key];
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(
      `The For you page has no stored wording for "${key}". The backend and ` +
        `this build disagree about the wording store; report it to the site ` +
        `administrator.`,
    );
  }
  return value;
}

/** Throw a sentence naming the failed act, or return the decoded body. */
async function orThrow<T>(response: Response, what: string): Promise<T> {
  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(`${what} (HTTP ${response.status}${detail}).`);
  }
  return (await response.json()) as T;
}

/**
 * The page, in one request.
 *
 * The shape check is deliberate and narrow: an absent `wording` would make
 * every label on the page throw one by one, and an absent `unread` would render
 * as "nothing waiting" — a confident, false screen. Both are contract
 * mismatches and both say so here, once.
 */
export async function fetchForYou(slug: string): Promise<ForYouPage> {
  const what = "The For you list could not be loaded";
  const response = await authFetch(
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}/for-you`,
    { timeoutMs: PRACTICE_TIMEOUT_MS },
  );
  const body = await orThrow<Partial<ForYouPage>>(response, what);
  if (
    typeof body.side !== "string" ||
    !Array.isArray(body.unread) ||
    !Array.isArray(body.everything) ||
    typeof body.wording !== "object" ||
    body.wording === null
  ) {
    throw new Error(`${what}: the response is not a For you page — backend/frontend contract mismatch.`);
  }
  return body as ForYouPage;
}

/** The badge's one number. Fetched once per app load and after a read-clear. */
export async function fetchForYouSummary(slug: string): Promise<ForYouSummary> {
  const what = "The For you count could not be read";
  const response = await authFetch(
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}/for-you/summary`,
    { timeoutMs: PRACTICE_TIMEOUT_MS },
  );
  const body = await orThrow<Partial<ForYouSummary>>(response, what);
  if (typeof body.unread_count !== "number" || typeof body.side !== "string") {
    throw new Error(`${what}: the response carries no count — backend/frontend contract mismatch.`);
  }
  return body as ForYouSummary;
}

/**
 * Opening a question marks everything on it read, for the signed-in person.
 *
 * Returns how many rows were newly written — 0 when she had already read it,
 * which is a legitimate state and why this is a number rather than a boolean.
 */
export async function markQuestionSeen(questionId: string): Promise<number> {
  const what = "The question could not be marked as read";
  const response = await authFetch(
    `${API_BASE_URL}/api/practice/questions/${encodeURIComponent(questionId)}/seen`,
    { method: "POST", timeoutMs: PRACTICE_TIMEOUT_MS },
  );
  const body = await orThrow<{ marked?: number }>(response, what);
  if (typeof body.marked !== "number") {
    throw new Error(`${what}: the response carries no count — backend/frontend contract mismatch.`);
  }
  return body.marked;
}

File: backend/src/dto/for_you.rs
//! The "For you" page, as the browser receives it (CC_TASK_FOR_YOU_v1 L1).
//!
//! ## Domain note: every sentence arrives finished
//!
//! A row carries its deck line, its body and its byline as COMPOSED strings,
//! and the two tab labels arrive with their counts already in them. The browser
//! holds no templates and fills no placeholder — the law `wording_practice_row`
//! states for the deck row, applied to a page whose whole content is sentences.
//! What that buys: retuning how a row reads is a Settings edit and a restart.
//!
//! The wording block rides along anyway, because the page has its own furniture
//! to render — three day headings, an empty state, a title — and it is the same
//! one-request shape the practice deck uses.

use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::for_you_wording::ForYouWordingDto;

/// Whose list this viewer is being served.
///
/// ## Rust Learning: `#[serde(rename_all = "snake_case")]` on a unit enum
///
/// The variants cross the wire as `"reviewers"`, `"witness"` and `"none"`
/// rather than as Rust's `Reviewers`. The browser compares strings, and a
/// capital letter in one of them would be a bug nothing in either language
/// could catch — so the casing is declared once, here.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ForYouSide {
    /// A listed reviewer, or an administrator (`may_review`).
    Reviewers,
    /// The witness.
    Witness,
    /// Neither. An honest empty page, never an error.
    None,
}

/// Which day heading a row belongs under.
///
/// Decided on the SERVER, in the case's own timezone. A browser in another
/// timezone would group the same rows differently from the deck they came out
/// of, and the person reading would have no way to tell which was right.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ForYouDay {
    Today,
    Yesterday,
    Earlier,
}

/// One row on the page.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForYouRowDto {
    /// `answer`, `note` or `change` — what the row is about. Carried so the
    /// page can mark the three apart visually; the SENTENCE that names the kind
    /// is already in `byline`.
    pub kind: String,
    /// The answer, note or change this row stands for.
    pub item_id: Uuid,
    pub scenario_id: Uuid,
    /// `None` only for a note about a whole scenario, which opens the deck
    /// rather than a question.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub question_id: Option<Uuid>,
    /// `S-11 · The $50,000 — “Marie, whose money was the $50,000 check…”`
    pub deck_line: String,
    /// The note's text, or the answer / new wording under its stored template.
    pub body: String,
    /// `Chuck · on your answer of Mon 21 Sep`
    pub byline: String,
    /// The time, in the case's timezone: the clock alone for today, the full
    /// stamp for anything older.
    pub when: String,
    pub day: ForYouDay,
    /// True on the Everything tab for a row this person has already read.
    pub read: bool,
}

/// The whole page, in one request.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForYouPayload {
    pub side: ForYouSide,
    /// The line under the title, with the other side already named.
    pub subtitle: String,
    /// `Unread · 3` — composed, so the browser fills no template.
    pub tab_unread_label: String,
    /// `Everything · 14`
    pub tab_everything_label: String,
    /// The badge's number. Serialized even at zero: the BROWSER decides not to
    /// draw a badge, and a missing field would be indistinguishable from a
    /// payload that forgot to carry it.
    pub unread_count: u32,
    pub unread: Vec<ForYouRowDto>,
    pub everything: Vec<ForYouRowDto>,
    /// The empty state's second line, with the other side named.
    pub empty_hint: String,
    /// Its third line. `None` when this side has never had an item — the line
    /// is then withheld entirely rather than rendered with an empty date.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub empty_last: Option<String>,
    pub wording: ForYouWordingDto,
}

/// The menu badge's one number, on its own address.
///
/// Its own endpoint rather than the whole page, because it is fetched on EVERY
/// page: the badge is the thing that carries the page to the rest of the app,
/// and making every screen pay for a list nobody is looking at would be the
/// wrong trade.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForYouSummaryDto {
    pub side: ForYouSide,
    /// Unread items across every deck of the case. Zero is a number, not an
    /// absent field — the badge's absence is the browser's decision.
    pub unread_count: u32,
}

/// What the read-clear write reports back.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct QuestionSeenResponse {
    pub question_id: Uuid,
    /// How many seen rows this call actually WROTE — zero when the person had
    /// already read everything on the question, which is a legitimate state and
    /// is reported as a number rather than swallowed.
    ///
    /// The badge is NOT recomputed here. This route is addressed by a question
    /// id alone (like every other question write), so it does not know which
    /// case the caller is looking at, and guessing would be worse than the one
    /// extra request the browser already makes to refresh the count.
    pub marked: u32,
}
